go/arrow/csv/reader.go

```diff
@@ -224,142 +246,185 @@ func (r *Reader) validate(recs []string) {
 	}
 }
 
-func (r *Reader) read(recs []string) {
-	for i, str := range recs {
-		switch r.schema.Field(i).Type.(type) {
-		case *arrow.BooleanType:
-			var v bool
-			switch str {
-			case "false", "False", "0":
-				v = false
-			case "true", "True", "1":
-				v = true
-			}
-			r.bld.Field(i).(*array.BooleanBuilder).Append(v)
-		case *arrow.Int8Type:
-			v := r.readI8(str)
-			r.bld.Field(i).(*array.Int8Builder).Append(v)
-		case *arrow.Int16Type:
-			v := r.readI16(str)
-			r.bld.Field(i).(*array.Int16Builder).Append(v)
-		case *arrow.Int32Type:
-			v := r.readI32(str)
-			r.bld.Field(i).(*array.Int32Builder).Append(v)
-		case *arrow.Int64Type:
-			v := r.readI64(str)
-			r.bld.Field(i).(*array.Int64Builder).Append(v)
-		case *arrow.Uint8Type:
-			v := r.readU8(str)
-			r.bld.Field(i).(*array.Uint8Builder).Append(v)
-		case *arrow.Uint16Type:
-			v := r.readU16(str)
-			r.bld.Field(i).(*array.Uint16Builder).Append(v)
-		case *arrow.Uint32Type:
-			v := r.readU32(str)
-			r.bld.Field(i).(*array.Uint32Builder).Append(v)
-		case *arrow.Uint64Type:
-			v := r.readU64(str)
-			r.bld.Field(i).(*array.Uint64Builder).Append(v)
-		case *arrow.Float32Type:
-			v := r.readF32(str)
-			r.bld.Field(i).(*array.Float32Builder).Append(v)
-		case *arrow.Float64Type:
-			v := r.readF64(str)
-			r.bld.Field(i).(*array.Float64Builder).Append(v)
-		case *arrow.StringType:
-			r.bld.Field(i).(*array.StringBuilder).Append(str)
+func (r *Reader) isNull(val string) bool {
+	for _, ns := range r.nullValues {
+		if val == ns {
+			return true
 		}
 	}
+	return false
 }
 
-func (r *Reader) readI8(str string) int8 {
-	v, err := strconv.ParseInt(str, 10, 8)
-	if err != nil && r.err == nil {
-		r.err = err
-		return 0
+func (r *Reader) read(recs []string) {
+	for i, str := range recs {
+		r.fieldConverter[i](r.bld.Field(i), str)
 	}
-	return int8(v)
 }
 
-func (r *Reader) readI16(str string) int16 {
-	v, err := strconv.ParseInt(str, 10, 16)
-	if err != nil && r.err == nil {
-		r.err = err
-		return 0
+func (r *Reader) initFieldConverter(field *arrow.Field) func(array.Builder, string) {
+	switch field.Type.(type) {
+	case *arrow.BooleanType:
+		return func(field array.Builder, str string) {
+			r.parseBool(field, str)
+		}
+	case *arrow.Int8Type:
+		return func(field array.Builder, str string) {
+			v, null := r.parseInt(field, str, 8)
```

Comment:
couldn't we have `parseI8`, `parseI16`, ... `parseU8`, ... `parseU64` like we had with the `readIxx` and `readUxx` methods?

this would remove the need for the `(value, isnull)` return tuple, like for `parseBool`.

Reply:
I believe I did this with slightly longer names.